server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchAction.java

```diff
@@ -172,6 +163,69 @@ private SearchData load(SearchWsRequest request) {
+    if (request.getProjectKey() == null) {
+      return null;
+    }
+
+    ComponentDto project = componentFinder.getByKey(dbSession, request.getProjectKey());
+    if (!project.getOrganizationUuid().equals(organization.getUuid())) {
+      throw new NotFoundException(format("Component key '%s' not found", project.getKey()));
+    }
+
+    return project;
+  }
+
+  private List<QProfileDto> searchProfiles(DbSession dbSession, SearchWsRequest request, OrganizationDto organization, List<QProfileDto> defaultProfiles,
+    @Nullable ComponentDto project) {
+    Collection<QProfileDto> profiles = selectAllProfiles(dbSession, organization);
+
+    return profiles.stream()
+      .filter(hasLanguagePlugin())
+      .filter(byLanguage(request))
+      .filter(byName(request))
+      .filter(byDefault(request, defaultProfiles))
+      .filter(byProject(dbSession, project, defaultProfiles))
+      .sorted(Q_PROFILE_COMPARATOR)
+      .collect(Collectors.toList());
+  }
+
+  private Predicate<QProfileDto> hasLanguagePlugin() {
+    return p -> languages.get(p.getLanguage()) != null;
+  }
+
+  private static Predicate<QProfileDto> byName(SearchWsRequest request) {
+    return p -> request.getProfileName() == null || Objects.equals(p.getName(), request.getProfileName());
+  }
+
+  private static Predicate<QProfileDto> byLanguage(SearchWsRequest request) {
+    return p -> request.getLanguage() == null || Objects.equals(p.getLanguage(), request.getLanguage());
+  }
+
+  private static Predicate<QProfileDto> byDefault(SearchWsRequest request, List<QProfileDto> defaultProfiles) {
+    Set<String> defaultProfileUuids = defaultProfiles.stream().map(QProfileDto::getKee).collect(Collectors.toSet());
+    return p -> !request.getDefaults() || defaultProfileUuids.contains(p.getKee());
+  }
+
+  private Predicate<QProfileDto> byProject(DbSession dbSession, @Nullable ComponentDto project, List<QProfileDto> defaultProfiles) {
+    if (project == null) {
+      return p -> true;
+    }
+    Map<String, QProfileDto> effectiveProfiles = defaultProfiles.stream().collect(Collectors.toMap(QProfileDto::getLanguage, identity()));
+    effectiveProfiles.putAll(dbClient.qualityProfileDao().selectAssociatedToProjectUuidAndLanguages(dbSession, project, getLanguageKeys()).stream()
+      .collect(MoreCollectors.uniqueIndex(QProfileDto::getLanguage)));
+    return p -> Objects.equals(p.getKee(), effectiveProfiles.get(p.getLanguage()).getKee());
+  }
+
+  private Collection<QProfileDto> selectAllProfiles(DbSession dbSession, OrganizationDto organization) {
+    return dbClient.qualityProfileDao().selectOrderedByOrganizationUuid(dbSession, organization);
+  }
+
+  private Set<String> getLanguageKeys() {
+    return Arrays.stream(languages.all()).map(Language::getKey).collect(MoreCollectors.toSet());
+  }
+
   private ComponentDto getProject(String moduleKey, DbSession dbSession) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this unused private "getProject" method. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUnusedPrivateMethod)